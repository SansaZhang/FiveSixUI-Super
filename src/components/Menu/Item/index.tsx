import * as React from "react";

import { MenuProps, ClickParams, SelectParams } from '..';
import { getClassName } from '../../_util/common';

export interface ItemProps extends MenuProps {
  itemKey: string;
  disabled?: boolean;
  keyPath?: Array<string>;
}

export default class Item extends React.Component<ItemProps, any> {
  constructor(props) {
    super(props);
  }

  getSelectedStatus(key: string, keysArray: Array<string> | undefined) {
    return keysArray ? keysArray.indexOf(key) > -1 : false;
  }
  handleClick(e) {
    const { disabled, onClick, itemKey, onSelect, selectedKeys, multiple } = this.props;
    if (disabled) return;

    const newKeyPath = this.addKeyPath(this.props);
    const clickParams: ClickParams = {
      key: itemKey,
      keyPath: newKeyPath,
      item: this,
      domEvent: e
    }

    const selectParams: SelectParams = {
      key: itemKey,
      keyPath: newKeyPath,
      item: this,
      domEvent: e,
      selectedKeys: multiple ? (selectedKeys || []).concat([itemKey]) : [itemKey]
    }

    onClick && onClick(clickParams);
    onSelect && onSelect(selectParams);
  }
  addKeyPath(info: ItemProps) {
    return (info.keyPath || []).concat(info.itemKey);
  }
  render() {
    const { selectedKeys, itemKey, className, disabled, onClick } = this.props;
    const selected = this.getSelectedStatus(itemKey, selectedKeys);

    return (
      <li 
        className={getClassName(`${className}-item`, {
          selected,
          disabled
        })}
        onClick={ (e) => this.handleClick(e) }
      >
        { this.props.children }
      </li>
    )
  }
}
