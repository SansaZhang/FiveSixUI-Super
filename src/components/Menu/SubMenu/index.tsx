import * as React from "react";
import { MenuProps, ClickParams } from '..';
import { getClassName, getKeyFromNode } from '../../_util/common';

export interface SubMenuProps extends MenuProps {
  itemKey: string;
  disabled?: boolean;
  title: string;
  keyPath?: Array<string>;
}

export default class SubMenu extends React.Component<SubMenuProps, any> {
  constructor(props: SubMenuProps) {
    super(props);
  }
  /**
   * 更倾向于下游来进行处理，要写更可信的下游
   * @param key 
   * @param keysArray 
   */
  getOpenStatus(key: string, keysArray: Array<string> | undefined) {
    if (keysArray) {
      for (let i = 0; i < keysArray.length; i++) {
        if (key == keysArray[i]) {
          return true;
        }
      }
    }
    return false;
  }
  componentWillReceiveProps(nextProps: SubMenuProps) {
    
  }
  handleClick(open:boolean) {
    const { disabled, onOpenChange, itemKey, openKeys } = this.props;
    if (disabled) return;

    const oldOpenKeys = openKeys || [];
    let newOpenKeys = oldOpenKeys.concat();
    if (open) { // 打开
      newOpenKeys = newOpenKeys.concat([itemKey]);
    } else { // 关闭
      const index = oldOpenKeys.indexOf(itemKey);
      index > -1 && newOpenKeys.splice(index, 1);
    }

    onOpenChange && onOpenChange(newOpenKeys);
  }

  addKeyPath(info: SubMenuProps) {
    return (info.keyPath || []).concat(info.itemKey);
  }
  render() {
    const { selectedKeys, itemKey, className, disabled, onClick, onSelect, onOpenChange, title, mode, openKeys, multiple } = this.props;

    const props = {
      className,
      selectedKeys,
      onClick,
      onSelect,
      onOpenChange,
      mode,
      openKeys,
      multiple,
      keyPath: this.addKeyPath(this.props)
    }
    const basicClassName = `${className}-submenu`;
    const open = this.getOpenStatus(itemKey, openKeys);

    return (
      <li 
        className={ getClassName(basicClassName, {
          disabled,
          selected: false,
          inline: mode === 'inline',
        }) }
      >
        <div 
          className={`${basicClassName}-title`}
          onClick={ () => this.handleClick(!open) }
        >
          { title }
        </div>
        { !open ? '' : 
          <ul className={ getClassName( className, { sub: true}) }>
            {
              React.Children.map(this.props.children, (child: React.ReactElement<any>, index) => {
                return React.cloneElement(child, {
                  ...props,
                  itemKey: getKeyFromNode(child, `${className}-item`, index)
                })
              })
            }
          </ul>
        }
      </li>
    )
  }
}
