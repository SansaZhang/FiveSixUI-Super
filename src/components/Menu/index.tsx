import * as React from "react";
import * as ReactDOM from 'react-dom';

import Item from './Item';
import SubMenu from './SubMenu';
import warning from '../_util/warning';
import { getClassName, getKeyFromNode } from '../_util/common';

/**
 * 【命名标准】组件名+Props
 */
export interface MenuProps {
  id?: string;
  defaultSelectedKeys?: Array<string>;// 当selected为undefined时才有效；
  selectedKeys?: Array<string>;
  defaultOpenKeys?: Array<string>;
  openKeys?: Array<string>;
  className?: string;
  prefixCls?: string;
  onClick?: (param: ClickParams) => void;
  onSelect?: (param: SelectParams) => void;
  onOpenChange?: (openKeys: Array<string>) => void;
  mode?: 'vertical' | 'horizontal' | 'inline'; // 暂不支持
  multiple?: boolean; // 暂不支持
  theme?: 'light' | 'dark'; // 暂不支持
}
/**
 * 【命名标准】组件名+State
 */
export interface MenuState {
  selectedKeys?: Array<string>;
  openKeys?: Array<string>;
}

export interface ClickParams {
  key: string;
  keyPath: Array<string>;
  item: any;
  domEvent: any;
}

export interface SelectParams {
  key: string;
  keyPath: Array<string>;
  item: any;
  domEvent: any;
  selectedKeys: Array<string>;
}

export default class Menu extends React.Component<MenuProps, MenuState> {
  static Item = Item;
  static SubMenu = SubMenu;
  static defaultProps = {
    prefixCls: 'wl-menu',
    className: '',
    theme: 'light',
    mode: 'inline'
  }
  /**
   * 所有函数输入，均需显式定义传入类型
   * @param props 
   */
  constructor(props: MenuProps) {
    super(props);

    warning(
      !('onOpen' in props),
      '`onOpen` is removed, please use `onOpenChange` instead'
    );

    const state: MenuState = {
      openKeys: props.openKeys || props.defaultOpenKeys || [],
      selectedKeys: props.selectedKeys || props.defaultSelectedKeys || []
    }
    this.state = state;
  }

  componentWillReceiveProps(nextProps: MenuProps) {
    let state: MenuState = {};
    if ('selectedKeys' in nextProps) {
      state.selectedKeys = nextProps.selectedKeys || [];
    }
    if ('openKeys' in nextProps) {
      state.openKeys = nextProps.openKeys || [];
    }
    this.setState(state);
  }

  handleClick = (params: ClickParams) => {
    const { onClick } = this.props;
    onClick && onClick(params)
  }

  handleSelect = (param: SelectParams) => {
    const { onSelect } = this.props;
    const selectedKeys = this.state.selectedKeys;

    if (selectedKeys && selectedKeys.indexOf(param.key) > -1) {
      return;
    }

    this.setState({
      selectedKeys: param.selectedKeys
    });

    onSelect && onSelect(param);
  }

  handleOpen = (openKeys: Array<string>) => {
    const { onOpenChange } = this.props;

    this.setState({
      openKeys
    });
    
    onOpenChange && onOpenChange(openKeys);
  }

  render() {
    // 注意这里className的设置 能够允许用户设置className
    const className: string = `${this.props.className} ${this.props.prefixCls}`;
    const { mode, multiple, prefixCls } = this.props;

    const props = {
      className: prefixCls,
      mode,
      multiple,
      selectedKeys: this.state.selectedKeys,
      openKeys: this.state.openKeys,
      onClick: this.handleClick,
      onSelect: this.handleSelect,
      onOpenChange: this.handleOpen
    }
    return (
      <ul className={ getClassName( className, {
        root: true,
        inline: mode === 'inline'
      })}>
        {
          React.Children.map(this.props.children, (child: React.ReactElement<any>, index) => {
            return React.cloneElement(child, {
              ...props,
              itemKey: getKeyFromNode(child, `${className}-item`, index)
            })
          })
        }
      </ul>
    )
  }
}
