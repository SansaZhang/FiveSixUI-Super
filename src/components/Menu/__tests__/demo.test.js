import React from 'react';
import Menu from '..';
import renderer from 'react-test-renderer';

// 简陋版GUI测试
// 在最开始需要撰写简单的应用场景，在开发、迭代时会提示对界面的更改，以确保更改是符合预期的

const SubMenu = Menu.SubMenu;

test('Menu demo', () => {
  const component = renderer.create(
    <Menu selectedKeys={['item3']} defaultSelectedKeys={['item1']} defaultOpenKeys={['subMenu']}>
      <Menu.Item key="item1">item1</Menu.Item>
      <Menu.Item key="item2">item2</Menu.Item>
    </Menu>
  );
  let tree = component.toJSON();
  expect(tree).toMatchSnapshot();
});

test('Menu demo', () => {
  const component = renderer.create(
    <Menu selectedKeys={['item3']} defaultSelectedKeys={['item1']} defaultOpenKeys={['subMenu']}>
      <Menu.Item key="item1">item1</Menu.Item>
      <Menu.Item key="item2">item2</Menu.Item>
      <SubMenu title="subMenu" key="subMenu">
        <Menu.Item key="item3">item3</Menu.Item>
      </SubMenu>
    </Menu>
  );
  let tree = component.toJSON();
  expect(tree).toMatchSnapshot();
});

test('Menu demo', () => {
  const component = renderer.create(
    <Menu selectedKeys={['item3']} defaultSelectedKeys={['item1']} defaultOpenKeys={['subMenu']}>
      <Menu.Item key="item1">item1</Menu.Item>
      <Menu.Item key="item2">item2</Menu.Item>
      <SubMenu title="subMenu" key="subMenu">
        <Menu.Item key="item3">item3</Menu.Item>
        <SubMenu title="subMenu2" key="subMenu2">
          <Menu.Item key="item32">item32</Menu.Item>
          <SubMenu title="subMenu3" key="subMenu3">
            <SubMenu title="subMenu4" key="subMenu4">
              <Menu.Item key="item34">item34</Menu.Item>
              <Menu.Item key="item35">item35</Menu.Item>
            </SubMenu>
            <Menu.Item key="item33">item33</Menu.Item>
          </SubMenu>
        </SubMenu>
      </SubMenu>
    </Menu>
  );
  let tree = component.toJSON();
  expect(tree).toMatchSnapshot();
});