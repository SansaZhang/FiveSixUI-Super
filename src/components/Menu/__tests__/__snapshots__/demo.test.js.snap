exports[`test Menu demo 1`] = `
<ul
  className=" wl-menu  wl-menu-root  wl-menu-inline">
  <li
    className="wl-menu-item"
    onClick={[Function]}>
    item1
  </li>
  <li
    className="wl-menu-item"
    onClick={[Function]}>
    item2
  </li>
</ul>
`;

exports[`test Menu demo 2`] = `
<ul
  className=" wl-menu  wl-menu-root  wl-menu-inline">
  <li
    className="wl-menu-item"
    onClick={[Function]}>
    item1
  </li>
  <li
    className="wl-menu-item"
    onClick={[Function]}>
    item2
  </li>
  <li
    className="wl-menu-submenu wl-menu-submenu-inline">
    <div
      className="wl-menu-submenu-title"
      onClick={[Function]}>
      subMenu
    </div>
    <ul
      className="wl-menu wl-menu-sub">
      <li
        className="wl-menu-item wl-menu-item-selected"
        onClick={[Function]}>
        item3
      </li>
    </ul>
  </li>
</ul>
`;

exports[`test Menu demo 3`] = `
<ul
  className=" wl-menu  wl-menu-root  wl-menu-inline">
  <li
    className="wl-menu-item"
    onClick={[Function]}>
    item1
  </li>
  <li
    className="wl-menu-item"
    onClick={[Function]}>
    item2
  </li>
  <li
    className="wl-menu-submenu wl-menu-submenu-inline">
    <div
      className="wl-menu-submenu-title"
      onClick={[Function]}>
      subMenu
    </div>
    <ul
      className="wl-menu wl-menu-sub">
      <li
        className="wl-menu-item wl-menu-item-selected"
        onClick={[Function]}>
        item3
      </li>
      <li
        className="wl-menu-submenu wl-menu-submenu-inline">
        <div
          className="wl-menu-submenu-title"
          onClick={[Function]}>
          subMenu2
        </div>
        
      </li>
    </ul>
  </li>
</ul>
`;
